Crash on opening

All Lexica v2 pre-release crash when I try to open the app. 
Lexica v1.1.0 (latest release) work properly. 
I'm running Xiaomi No Note 10 Lite on Android 10 

The error log: 

```
java.lang.NullPointerException: Parameter specified as non-null is null: method kotlin.jvm.internal.Intrinsics.checkNotNullParameter, parameter language
	at com.serwylo.lexica.db.ResultRepository.findHighScore(Unknown Source:7)
	at com.serwylo.lexica.MainMenuActivity.lambda$load$7$MainMenuActivity(MainMenuActivity.java:126)
	at com.serwylo.lexica.-$$Lambda$MainMenuActivity$034F0fTih04RzSV0E_Hm0ZJa6G8.run(Unknown Source:2)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:919)
```

I have something new : when I update Lexica from v1, v2 work. But the app crash when i install directly v2 on my phone.
My log :

                            Process com.serwylo.lexica created for activity {com.serwylo.lexica/com.serwylo.lexica.MainMenuActivity} caller=com.miui.home
                            PID: 31257   UID:    GIDs: 

         .serwylo.lexic  E  Not starting debugger since process cannot load the jdwp agent.
                   Perf  I  Connecting to perf service.
          FeatureParser  I  can't find toco.xml in assets/device_features/,it may be in /vendor/etc/device_features
        .serwylo.lexica  W  type=1400 audit(0.0:101591): avc: denied { read } for name="u:object_r:vendor_displayfeature_prop:s0" dev="tmpfs" ino=20033 scontext=u:r:untrusted_app_27:s0:c106,c257,c512,c768 tcontext=u:object_r:vendor_displayfeature_prop:s0 tclass=file permissive=0
                   libc  E  Access denied finding property "ro.vendor.df.effect.conflict"
                   Perf  E  Fail to get file list com.serwylo.lexica
                         E  getFolderSize() : Exception_1 = java.lang.NullPointerException: Attempt to get length of null array
                         E  Fail to get file list com.serwylo.lexica
                         E  getFolderSize() : Exception_1 = java.lang.NullPointerException: Attempt to get length of null array
                         E  Fail to get file list oat
                         E  getFolderSize() : Exception_1 = java.lang.NullPointerException: Attempt to get length of null array
        ForceDarkHelper  D  updateByCheckExcludeList: pkg: com.serwylo.lexica activity: com.serwylo.lexica.MainMenuActivity@adbba76
                   Util  D  Searching for lexicon that best matches the locale: "fr_FR"
                         D  Found matching language code "fr". Will now check country code.
                         D  Remembering language "fr" in case we don't find any better match.
        ForceDarkHelper  D  updateByCheckExcludeList: pkg: com.serwylo.lexica activity: com.serwylo.lexica.MainMenuActivity@adbba76
                   Util  D  No exact match, so the best option is matching language code with no specific country: "fr"
     GameModeRepository  I  No current game mode stored (perhaps first run?). Choosing the first game mode from the list of modes.
                         I  Remembering the current game mode: Game Mode [id: 1, type: SPRINT]
        ForceDarkHelper  D  updateByCheckExcludeList: pkg: com.serwylo.lexica activity: com.serwylo.lexica.MainMenuActivity@adbba76
         AndroidRuntime  E  FATAL EXCEPTION: AsyncTask #1
                         E  Process: com.serwylo.lexica, PID: 31257
                         E  java.lang.NullPointerException: Parameter specified as non-null is null: method kotlin.jvm.internal.Intrinsics.checkNotNullParameter, parameter language
                         E      at com.serwylo.lexica.db.ResultRepository.findHighScore(Unknown Source:7)
                         E      at com.serwylo.lexica.MainMenuActivity.lambda$load$7$MainMenuActivity(MainMenuActivity.java:126)
                         E      at com.serwylo.lexica.-$$Lambda$MainMenuActivity$034F0fTih04RzSV0E_Hm0ZJa6G8.run(Unknown Source:2)
                         E      at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
                         E      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
                         E      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
                         E      at java.lang.Thread.run(Thread.java:919)
                 chatty  I  uid=10362(com.serwylo.lexica) .serwylo.lexica identical 1 line
        ForceDarkHelper  D  updateByCheckExcludeList: pkg: com.serwylo.lexica activity: com.serwylo.lexica.MainMenuActivity@adbba76
                   Util  D  Searching for lexicon that best matches the locale: "fr_FR"
                         D  Found matching language code "fr". Will now check country code.
                         D  Remembering language "fr" in case we don't find any better match.
                         D  No exact match, so the best option is matching language code with no specific country: "fr"
                Process  I  Sendig signal. PID: 31257 SIG: 9

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -10,8 +10,8 @@ android {
         applicationId "com.serwylo.lexica"
         minSdkVersion 18
         targetSdkVersion 28
-        versionCode 20001
-        versionName "2.0.0-alpha2"
+        versionCode 20002
+        versionName "2.0.0-alpha3"
 
         testInstrumentationRunner 'androidx.test.runner.AndroidJUnitRunner'
     }

--- a/app/src/main/java/com/serwylo/lexica/ChooseLexiconActivity.kt
+++ b/app/src/main/java/com/serwylo/lexica/ChooseLexiconActivity.kt
@@ -36,7 +36,7 @@ class ChooseLexiconActivity : AppCompatActivity() {
     inner class Adapter : RecyclerView.Adapter<ViewHolder>() {
 
         private val languages: List<Language> = LanguageLabel.getAllLanguagesSorted(this@ChooseLexiconActivity)
-        private val selectedLanguage: Language?
+        private val selectedLanguage: Language = Util().getSelectedLanguageOrDefault(this@ChooseLexiconActivity)
 
         override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
             val binding = LexiconListItemBinding.inflate(this@ChooseLexiconActivity.layoutInflater, parent, false)
@@ -45,18 +45,13 @@ class ChooseLexiconActivity : AppCompatActivity() {
 
         override fun onBindViewHolder(holder: ViewHolder, position: Int) {
             val language = languages[position]
-            holder.bind(language, language.name == selectedLanguage?.name) { selectLexicon(language) }
+            holder.bind(language, language.name == selectedLanguage.name) { selectLexicon(language) }
         }
 
         override fun getItemCount(): Int {
             return languages.size
         }
 
-        init {
-
-            val languageCode = Util().getLexiconString(this@ChooseLexiconActivity)
-            selectedLanguage = Language.fromOrNull(languageCode)
-        }
     }
 
     private fun selectLexicon(language: Language) {

--- a/app/src/main/java/com/serwylo/lexica/MainMenuActivity.java
+++ b/app/src/main/java/com/serwylo/lexica/MainMenuActivity.java
@@ -66,8 +66,7 @@ public class MainMenuActivity extends AppCompatActivity {
         binding.gameModeButton.setOnClickListener(v -> startActivity(new Intent(this, ChooseGameModeActivity.class)));
         binding.gameModeButton.setText(gameMode.label(this));
 
-        String languageCode = new Util().getLexiconString(this);
-        Language language = Language.fromOrNull(languageCode);
+        Language language = new Util().getSelectedLanguageOrDefault(this);
         binding.languageButton.setText(LanguageLabel.getLabel(this, language));
         binding.languageButton.setOnClickListener(v -> startActivity(new Intent(this, ChooseLexiconActivity.class)));
 
@@ -112,8 +111,7 @@ public class MainMenuActivity extends AppCompatActivity {
             final Database db = Database.get(this);
             db.getOpenHelper().getReadableDatabase();
 
-            String languageCode = new Util().getLexiconString(this);
-            Language language = Language.fromOrNull(languageCode);
+            Language language = new Util().getSelectedLanguageOrDefault(this);
 
             final GameModeRepository gameModeRepository = new GameModeRepository(db.gameModeDao(), PreferenceManager.getDefaultSharedPreferences(this));
             final ResultRepository resultRepository = new ResultRepository(db.resultDao());

--- a/app/src/main/java/com/serwylo/lexica/Util.java
+++ b/app/src/main/java/com/serwylo/lexica/Util.java
@@ -6,6 +6,8 @@ import android.util.Log;
 
 import androidx.preference.PreferenceManager;
 
+import com.serwylo.lexica.lang.EnglishGB;
+import com.serwylo.lexica.lang.EnglishUS;
 import com.serwylo.lexica.lang.Language;
 
 import java.util.Collection;
@@ -14,28 +16,59 @@ import java.util.Locale;
 public class Util {
     private static final String TAG = "Util";
 
-    public String getLexiconString(Context context) {
+    /**
+     * Return the language that the user last chose to play with.
+     * Alternatively, if they have not explicitly chosen to do so, then ask the system
+     * what it's locale is and try to find the most appropriate language based on this.
+     */
+    public Language getSelectedLanguageOrDefault(Context context) {
         // Default to the language explicitly chosen by the user
         SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
         String chosenLanguage = prefs.getString("dict", null);
 
         if (chosenLanguage != null) {
             Log.d(TAG, "User explicitly chose " + chosenLanguage);
-            return chosenLanguage;
+
+            // Legacy preferences, which use either "US" or "UK" rather than the locale name (i.e. "en_US" or "en_GB")
+            if ("UK".equals(chosenLanguage)) {
+                Language language = new EnglishGB();
+                Log.i(TAG, "Detected legacy language preference GB, Updating to " + language.getName());
+                saveLanguagePref(prefs, language);
+                return language;
+            } else if ("US".equals(chosenLanguage)) {
+                Language language = new EnglishUS();
+                Log.i(TAG, "Detected legacy language preference US, Updating to " + language.getName());
+                saveLanguagePref(prefs, language);
+                return language;
+            }
+
+            try {
+                return Language.from(chosenLanguage);
+            } catch (Language.NotFound notFound) {
+                Log.e(TAG, "The previously chosen language " + chosenLanguage + " could not be found. Defaulting to US", notFound);
+                Language defaultLang = new EnglishUS();
+                saveLanguagePref(prefs, defaultLang);
+                return defaultLang;
+            }
         }
 
         Locale systemLocale = context.getResources().getConfiguration().locale;
 
         Language bestEffort = findBestMatchOrNull(systemLocale, Language.getAllLanguages().values());
         if (bestEffort != null) {
-            return bestEffort.getLocale().toString();
+            return bestEffort;
         }
 
         // Default
         Log.d(TAG, "Could not detect language for system language " + systemLocale + ", defaulting to US");
-        return "US";
+        Language defaultLang = new EnglishUS();
+        saveLanguagePref(prefs, defaultLang);
+        return defaultLang;
     }
 
+    private static void saveLanguagePref(SharedPreferences prefs, Language language) {
+        prefs.edit().putString("dict", language.getName()).apply();
+    }
     /**
      * If we don't find an exact match, then we will still try to give the user a lexicon
      * from the same language, even if not the same country. In order of precedence, prefer:

--- a/app/src/main/java/com/serwylo/lexica/game/Game.java
+++ b/app/src/main/java/com/serwylo/lexica/game/Game.java
@@ -238,16 +238,7 @@ public class Game implements Synchronizer.Counter {
     private void loadPreferences(Context c, GameMode gameMode) {
         SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(c);
 
-        String languageCode = new Util().getLexiconString(context);
-        language = Language.fromOrNull(languageCode);
-        if (language == null) {
-            // Legacy preferences, which use either "US" or "UK" rather than the locale name (i.e. "en_US" or "en_GB")
-            if ("UK".equals(languageCode)) {
-                language = new EnglishGB();
-            } else {
-                language = new EnglishUS();
-            }
-        }
+        language = new Util().getSelectedLanguageOrDefault(context);
         Log.d(TAG, "Language (from preferences): " + language.getName());
 
         boardSize = gameMode.getBoardSize();

--- a/fastlane/metadata/android/en-US/changelogs/20003.txt
+++ b/fastlane/metadata/android/en-US/changelogs/20003.txt
@@ -1,0 +1,16 @@
+Game Modes!
+
+These preferences are replaced with "Game Modes":
+ * Duration
+ * Board size
+ * Min word length
+ * Scoring type
+ * Hint modes
+
+You can change game mode from the main screen, or create your own. High scores are based on game mode and lexicon.
+
+Fixed critical bug in 2.0.0-alpha2 for French and German locales (thanks to all who reported the issue).
+
+Please provide feedback/report issues at https://github.com/lexica/lexica/issues.
+
+Support Lexica development via Liberapay or GitHub Sponsors :)

--- a/libraries/trie/src/test/java/com/serwylo/lexica/lang/LanguageTest.kt
+++ b/libraries/trie/src/test/java/com/serwylo/lexica/lang/LanguageTest.kt
@@ -1,0 +1,26 @@
+package com.serwylo.lexica.lang
+
+import org.junit.Test
+
+import org.junit.Assert.*
+
+class LanguageTest {
+
+    /**
+     * Part of adding a new language means putting it in the [Language.allLanguages] map.
+     * This map requires a key to be present which matches the name of the language, as
+     * it will be used for lookups later. This ensures it is done correctly.
+     */
+    @Test
+    fun ensureLanguagesAreCorrectlyIndexed() {
+
+        Language.allLanguages.forEach {
+            val code = it.key
+            val lang = it.value
+
+            assertEquals(code, lang.name)
+        }
+
+    }
+
+}

